core/Pos/Binary/Class/Primitive.hs

```diff
@@ -3,110 +3,147 @@
+import           Control.Monad.ST              (ST, runST)
+import qualified Data.ByteString               as BS
+import qualified Data.ByteString.Lazy          as BSL
+import qualified Data.ByteString.Lazy.Internal as BSL
+import           Data.SafeCopy                 (Contained, SafeCopy (..), contain,
+                                                safeGet, safePut)
+import qualified Data.Serialize                as Cereal (Get, Put)
+import           Data.Typeable                 (typeRep)
+
+import           Pos.Binary.Class.Core         (Bi (..))
+import           Serokell.Data.Memory.Units    (Byte)
+import           Universum
 
-       -- * Serialization strings and sequence of bytes
-       , getBytes
-       , putBytes
+-- | Serialize a Haskell value to an external binary representation.
+--
+-- The output is represented as a lazy 'BSL.ByteString' and is constructed
+-- incrementally.
+serialize :: Bi a => a -> BSL.ByteString
+serialize = CBOR.Write.toLazyByteString . encode
 
-       , getAsciiString1b
-       , putAsciiString1b
-       , sizeAsciiString1b
+-- | Serialize a Haskell value to an external binary representation.
+--
+-- The output is represented as a strict 'BS.ByteString'.
+serialize' :: Bi a => a -> BS.ByteString
+serialize' = BSL.toStrict . serialize
 
-       , Raw (..)
-       ) where
+-- | Deserialize a Haskell value from the external binary representation
+-- (which must have been made using 'serialize' or related function).
+--
+-- /Throws/: @'CBOR.Read.DeserialiseFailure'@ if the given external
+-- representation is invalid or does not correspond to a value of the
+-- expected type.
+deserialize :: Bi a => BSL.ByteString -> a
+deserialize = either throw identity . bimap third third . deserializeOrFail
+    where
+      third :: (a, b, c) -> c
+      third (_, _, c) = c
+
+-- | Strict variant of 'deserialize'.
+deserialize' :: Bi a => BS.ByteString -> a
+deserialize' = deserialize . BSL.fromStrict
+
+-- | Deserialize a Haskell value from the external binary representation,
+-- failing if there are leftovers. In a nutshell, the `full` here implies
+-- the contract of this function is that what you feed as input needs to
+-- be consumed entirely.
+decodeFull :: Bi a => BS.ByteString -> Either Text a
+decodeFull bs0 = case deserializeOrFail' bs0 of
+  Right (leftover, offset, x) -> case BS.null leftover of
+      True  -> pure x
+      False ->
+          let msg = "decodeFull failed! Leftover found at offset " <> show offset <> ": " <> show leftover
+          in Left $ fromString msg
+  Left  e -> Left $ fromString (show e)
+
+-- | Deserialize a Haskell value from the external binary representation,
+-- returning either (leftover, offset, value) or a @'DeserialiseFailure'@.
+deserializeOrFail
+    :: Bi a
+    => BSL.ByteString
+    -> Either (BS.ByteString, CBOR.Read.ByteOffset, CBOR.Read.DeserialiseFailure)
+              (BS.ByteString, CBOR.Read.ByteOffset, a)
```

Comment:
> Don't bother returning the CBOR.Read.ByteOffset: it's never useful to us and in the current cborg release it is not even accurate (iirc it's always 0).

Perfect! I really had no clue if we were going to use it at some stage, so I thought about including it at first 😉  consider it done!